File: backend/opencode/thinking.go
package opencode

import (
	"strings"
)

// ThinkingLevel controls model reasoning effort (OpenCode + Codex + Responses API).
type ThinkingLevel string

const (
	ThinkingOff     ThinkingLevel = "off"
	ThinkingMinimal ThinkingLevel = "minimal"
	ThinkingLow     ThinkingLevel = "low"
	ThinkingMedium  ThinkingLevel = "medium"
	ThinkingHigh    ThinkingLevel = "high"
	ThinkingMax     ThinkingLevel = "max"
	ThinkingXHigh   ThinkingLevel = "xhigh"
)

type ThinkingParams struct {
	Type string `json:"type"`
}

// deepseekV4FlashLevels matches OpenCode's opencode-go deepseek-v4 efforts.
var deepseekV4FlashLevels = []ThinkingLevel{ThinkingLow, ThinkingMedium, ThinkingHigh, ThinkingMax}

// defaultReasoningLevels is the standard OpenAI/Codex reasoning effort ladder.
var defaultReasoningLevels = []ThinkingLevel{
	ThinkingOff, ThinkingMinimal, ThinkingLow, ThinkingMedium, ThinkingHigh, ThinkingXHigh,
}

func earlyReturnVariants(model string) bool {
	id := strings.ToLower(model)
	if strings.Contains(id, "minimax-m3") {
		return false
	}
	for _, part := range []string{
		"deepseek-chat", "deepseek-reasoner", "deepseek-r1", "deepseek-v3",
		"minimax", "glm", "kimi", "k2p", "qwen", "big-pickle",
	} {
		if strings.Contains(id, part) {
			return true
		}
	}
	return false
}

func SupportsThinking(model string) bool {
	return len(SupportedThinkingLevels(model)) > 1
}

func mandatoryThinking(model string) bool {
	return opencodeMandatoryThinkingID(model)
}

func SupportedThinkingLevels(model string) []ThinkingLevel {
	modelLower := strings.ToLower(model)
	if strings.Contains(modelLower, "gpt-") {
		return append([]ThinkingLevel(nil), defaultReasoningLevels...)
	}
	if strings.Contains(modelLower, "minimax-m3") {
		return []ThinkingLevel{ThinkingOff, ThinkingMedium}
	}
	if earlyReturnVariants(model) {
		if mandatoryThinking(model) {
			return []ThinkingLevel{ThinkingMedium}
		}
		return []ThinkingLevel{ThinkingOff}
	}
	if strings.Contains(modelLower, "deepseek-v4") {
		return append([]ThinkingLevel(nil), deepseekV4FlashLevels...)
	}
	return []ThinkingLevel{ThinkingLow, ThinkingMedium, ThinkingHigh}
}

func CycleThinkingLevel(current ThinkingLevel, model string) ThinkingLevel {
	levels := SupportedThinkingLevels(model)
	if len(levels) <= 1 {
		return ThinkingOff
	}
	idx := 0
	for i, l := range levels {
		if l == current {
			idx = i
			break
		}
	}
	return levels[(idx+1)%len(levels)]
}

func StepThinkingLevel(current ThinkingLevel, model string, delta int) ThinkingLevel {
	levels := SupportedThinkingLevels(model)
	if len(levels) <= 1 {
		return ThinkingOff
	}
	idx := 0
	for i, l := range levels {
		if l == current {
			idx = i
			break
		}
	}
	n := len(levels)
	idx = ((idx+delta)%n + n) % n
	return levels[idx]
}

func ApplyThinkingToRequest(req *ChatRequest, level ThinkingLevel, model string) {
	if !SupportsThinking(model) {
		return
	}
	modelLower := strings.ToLower(model)
	if strings.Contains(modelLower, "minimax-m3") {
		if level == ThinkingOff || level == "" {
			req.Thinking = &ThinkingParams{Type: "disabled"}
		} else {
			req.Thinking = &ThinkingParams{Type: "adaptive"}
		}
		req.ReasoningEffort = ""
		return
	}

	if level == ThinkingOff || level == "" {
		req.ReasoningEffort = ""
		req.Thinking = nil
		return
	}

	effort := ReasoningEffortForAPI(level, model)
	req.ReasoningEffort = effort
	req.Thinking = nil
}

// ReasoningEffortForAPI maps UI thinking levels to provider wire values.
func ReasoningEffortForAPI(level ThinkingLevel, model string) string {
	if level == ThinkingOff || level == "" {
		return ""
	}
	if strings.Contains(strings.ToLower(model), "deepseek-v4") {
		if level == ThinkingXHigh || level == ThinkingMax {
			return "max"
		}
		return string(level)
	}
	if level == ThinkingMax {
		return "xhigh"
	}
	return string(level)
}

// SupportsReasoning returns true if the model supports reasoning output.
func SupportsReasoning(model string) bool {
	if m, ok := LookupCatalogModel(model); ok {
		return m.Reasoning || m.ReasoningField != ""
	}
	// Fallback check before catalog is loaded/offline
	modelLower := strings.ToLower(model)
	if opencodeMandatoryThinkingID(model) {
		return true
	}
	if strings.Contains(modelLower, "deepseek-v4") || strings.Contains(modelLower, "deepseek-r1") || strings.Contains(modelLower, "reasoner") {
		return true
	}
	if strings.Contains(modelLower, "mimo") || strings.Contains(modelLower, "hy3") {
		return true
	}
	if strings.Contains(modelLower, "gpt-") {
		return true
	}
	return false
}

// NormalizeMessages ensures assistant messages include reasoning_content when required.
func NormalizeMessages(msgs []Message, model string) []Message {
	if !SupportsReasoning(model) {
		return msgs
	}
	field := "reasoning_content"
	if m, ok := LookupCatalogModel(model); ok && m.ReasoningField != "" {
		field = m.ReasoningField
	}

	out := make([]Message, len(msgs))
	copy(out, msgs)
	for i := range out {
		if out[i].Role != "assistant" {
			continue
		}
		reasoningText := out[i].GetReasoning()
		out[i].ReasoningContent = nil
		out[i].ReasoningDetails = nil
		out[i].ReasoningPlain = nil

		switch field {
		case "reasoning_details":
			out[i].ReasoningDetails = &reasoningText
		case "reasoning":
			out[i].ReasoningPlain = &reasoningText
		default:
			out[i].ReasoningContent = &reasoningText
		}
	}
	return out
}

func ParseThinkingLevel(s string) ThinkingLevel {
	switch ThinkingLevel(s) {
	case ThinkingMinimal, ThinkingLow, ThinkingMedium, ThinkingHigh, ThinkingMax, ThinkingXHigh:
		return ThinkingLevel(s)
	default:
		return ThinkingOff
	}
}

func FormatThinkingLabel(level ThinkingLevel) string {
	if level == "" || level == ThinkingOff {
		return "off"
	}
	return string(level)
}

// FormatThinkingLevelForModel returns the user-facing label for a thinking level.
func FormatThinkingLevelForModel(model string, level ThinkingLevel) string {
	if strings.Contains(strings.ToLower(model), "minimax-m3") {
		if level == ThinkingOff || level == "" {
			return "none"
		}
		return "thinking"
	}
	return FormatThinkingLabel(level)
}

File: backend/opencode/thinking.ts
import { get_reasoning, type chat_request, type message } from "./types";
import { lookup_catalog_model } from "./providers";

export type thinking_level_val = "off" | "adaptive" | "minimal" | "low" | "medium" | "high" | "max" | "xhigh" | "";

export const thinking_off: thinking_level_val = "off";
export const thinking_adaptive: thinking_level_val = "adaptive";
export const thinking_minimal: thinking_level_val = "minimal";
export const thinking_low: thinking_level_val = "low";
export const thinking_medium: thinking_level_val = "medium";
export const thinking_high: thinking_level_val = "high";
export const thinking_max: thinking_level_val = "max";
export const thinking_xhigh: thinking_level_val = "xhigh";

export type thinking_params = {
  type: string;
};

export const deepseek_v4_flash_levels: thinking_level_val[] = [
  "low",
  "medium",
  "high",
  "max"
];

export const default_reasoning_levels: thinking_level_val[] = [
  "off",
  "minimal",
  "low",
  "medium",
  "high",
  "xhigh"
];

export const early_return_variants = (model: string): boolean => {
  const id = model.toLowerCase();
  if (id.includes("minimax")) {
    return false;
  }
  const parts = [
    "deepseek-chat",
    "deepseek-reasoner",
    "deepseek-r1",
    "deepseek-v3",
    "glm",
    "kimi",
    "k2p",
    "qwen",
    "big-pickle"
  ];
  for (const part of parts) {
    if (id.includes(part)) {
      return true;
    }
  }
  return false;
};

export const supports_thinking = (model: string): boolean => {
  return supported_thinking_levels(model).length > 1;
};

export const mandatory_thinking = (model: string): boolean => {
  return opencode_mandatory_thinking_id(model);
};

export const opencode_mandatory_thinking_id = (id: string): boolean => {
  const lower_id = id.toLowerCase();
  if (
    lower_id.includes("deepseek-chat") ||
    lower_id.includes("deepseek-reasoner") ||
    lower_id.includes("deepseek-r1") ||
    lower_id.includes("deepseek-v3")
  ) {
    return true;
  }
  const parts = ["glm", "kimi", "k2p", "qwen", "big-pickle"];
  for (const part of parts) {
    if (lower_id.includes(part)) {
      return true;
    }
  }
  return false;
};

export const is_minimax_model = (model: string): boolean => model.toLowerCase().includes("minimax");

export const is_minimax_m3 = (model: string): boolean => model.toLowerCase().includes("minimax-m3");

export const supported_thinking_levels = (model: string): thinking_level_val[] => {
  const model_lower = model.toLowerCase();
  if (model_lower.includes("gpt-")) {
    return [...default_reasoning_levels];
  }
  if (is_minimax_m3(model_lower)) {
    // MiniMax-M3: thinking.type is "disabled" or "adaptive" — not reasoning_effort tiers.
    return ["off", "adaptive"];
  }
  if (is_minimax_model(model_lower)) {
    // M2.x: thinking is always on and cannot be disabled.
    return [];
  }
  if (early_return_variants(model)) {
    if (mandatory_thinking(model)) {
      return ["medium"];
    }
    return ["off"];
  }
  if (model_lower.includes("deepseek-v4")) {
    return [...deepseek_v4_flash_levels];
  }
  // Unknown models: do not assume thinking support. The resolver will still
  // flag well-known families above, and users can override via ~/.hollow/models.json.
  return [];
};

export const cycle_thinking_level = (current: thinking_level_val, model: string): thinking_level_val => {
  const levels = supported_thinking_levels(model);
  if (levels.length <= 1) {
    return "off";
  }
  let idx = 0;
  for (let i = 0; i < levels.length; i++) {
    if (levels[i] === current) {
      idx = i;
      break;
    }
  }
  return levels[(idx + 1) % levels.length];
};

export const step_thinking_level = (current: thinking_level_val, model: string, delta: number): thinking_level_val => {
  const levels = supported_thinking_levels(model);
  if (levels.length <= 1) {
    return "off";
  }
  let idx = 0;
  for (let i = 0; i < levels.length; i++) {
    if (levels[i] === current) {
      idx = i;
      break;
    }
  }
  const n = levels.length;
  idx = ((idx + delta) % n + n) % n;
  return levels[idx];
};

export const normalize_thinking_level = (level: string, model: string): thinking_level_val => {
  const parsed = parse_thinking_level(level);
  if (is_minimax_m3(model) && (parsed === "medium" || level === "thinking")) {
    return "adaptive";
  }
  return parsed;
};

export const default_thinking_level = (model: string): thinking_level_val => {
  const levels = supported_thinking_levels(model);
  if (levels.length === 0) {
    return "off";
  }
  if (levels.includes("adaptive")) {
    return "adaptive";
  }
  if (levels.includes("medium")) {
    return "medium";
  }
  return levels.find((l) => l !== "off") ?? levels[0];
};

const apply_minimax_thinking = (req: chat_request, level: thinking_level_val, model: string): void => {
  const normalized = normalize_thinking_level(level, model);
  if (is_minimax_m3(model)) {
    if (normalized === "off") {
      req.thinking = { type: "disabled" };
    } else {
      // Empty level means "use API default" (adaptive on); do not send disabled.
      req.thinking = { type: "adaptive" };
    }
  } else {
    req.thinking = { type: "adaptive" };
  }
  // MiniMax only streams reasoning_content / reasoning_details when this is set.
  req.reasoning_split = true;
  delete req.reasoning_effort;
};

export const apply_thinking_to_request = (req: chat_request, level: thinking_level_val, model: string): void => {
  if (is_minimax_model(model)) {
    apply_minimax_thinking(req, level, model);
    return;
  }
  if (!supports_thinking(model)) {
    return;
  }

  if (level === "off" || level === "") {
    delete req.reasoning_effort;
    delete req.thinking;
    return;
  }

  const effort = reasoning_effort_for_api(level, model);
  req.reasoning_effort = effort;
  delete req.thinking;
};

export const reasoning_effort_for_api = (level: thinking_level_val, model: string): string => {
  if (level === "off" || level === "") {
    return "";
  }
  if (model.toLowerCase().includes("deepseek-v4")) {
    if (level === "xhigh" || level === "max") {
      return "max";
    }
    return level;
  }
  if (level === "max") {
    return "xhigh";
  }
  return level;
};

export const supports_reasoning = (model: string): boolean => {
  const [m, ok] = lookup_catalog_model(model);
  if (ok) {
    return m.reasoning || (m.reasoning_field !== undefined && m.reasoning_field !== "");
  }
  const model_lower = model.toLowerCase();
  if (opencode_mandatory_thinking_id(model)) {
    return true;
  }
  if (
    model_lower.includes("deepseek-v4") ||
    model_lower.includes("deepseek-r1") ||
    model_lower.includes("reasoner")
  ) {
    return true;
  }
  if (model_lower.includes("mimo") || model_lower.includes("hy3")) {
    return true;
  }
  if (model_lower.includes("gpt-")) {
    return true;
  }
  if (is_minimax_model(model)) {
    return true;
  }
  return false;
};

export const normalize_messages = (msgs: message[], model: string): message[] => {
  if (!supports_reasoning(model)) {
    return msgs;
  }
  let field = "reasoning_content";
  const [m, ok] = lookup_catalog_model(model);
  if (ok && m.reasoning_field !== undefined && m.reasoning_field !== "") {
    field = m.reasoning_field;
  }

  return msgs.map((msg) => {
    if (msg.role !== "assistant") {
      return msg;
    }
    const reasoning_text = get_reasoning(msg);
    const next_msg = { ...msg };
    delete next_msg.reasoning_content;
    delete next_msg.reasoning_details;
    delete next_msg.reasoning;

    switch (field) {
      case "reasoning_details":
        next_msg.reasoning_details = reasoning_text;
        break;
      case "reasoning":
        next_msg.reasoning = reasoning_text;
        break;
      default:
        next_msg.reasoning_content = reasoning_text;
    }
    return next_msg;
  });
};

export const parse_thinking_level = (s: string): thinking_level_val => {
  switch (s) {
    case "adaptive":
    case "thinking":
      return "adaptive";
    case "minimal":
    case "low":
    case "medium":
    case "high":
    case "max":
    case "xhigh":
      return s;
    default:
      return "off";
  }
};

export const format_thinking_label = (level: thinking_level_val): string => {
  if (level === "" || level === "off") {
    return "off";
  }
  return level;
};

export const format_thinking_level_for_model = (model: string, level: thinking_level_val): string => {
  if (is_minimax_m3(model)) {
    if (level === "off" || level === "") {
      return "off";
    }
    if (level === "adaptive" || level === "medium") {
      return "thinking";
    }
  }
  return format_thinking_label(level);
};

